src/qibojit/custom_operators/__init__.py

```diff
@@ -93,8 +93,19 @@ def collapse_state(state, qubits, result, nqubits, normalize=True):
     return backend.collapse_state(state, qubits, result, nqubits, normalize)
 
 
+def transpose_state(pieces, state, nqubits, order):
+    # always fall back to numba CPU backend because for ops not implemented on GPU
+    numba_backend = backend.get("numba")
+    return numba_backend.transpose_state(pieces, state, nqubits, order)
```

Comment:
Just a quick question concerning this point. What happens if the state is allocated on GPU, does the code raises a proper warning or copy to CPU?

Reply:
If `state` is a cupy array it fails with the following message:
```
This error may have been caused by the following argument(s):
- argument 1: Cannot determine Numba type of <class 'cupy._core.core.ndarray'>
```
I believe this is the case for all numba ops, including the ones that apply gates. 

Would you prefer to automatically cast to CPU? If yes, should this be silent or with a warning?

Reply:
That's ok, if I understand correctly this specific configuration will never happen in practice, right?

Reply:
In principle no. Since the qibojit ops are only for "internal" usage in qibo, if we take proper care of which device each tensor is casted, we should not have such issues. In fact I believe that it is better to not do silent automatic castings from GPU to CPU as these typically reduce performance. If there is a device mismatch somewhere it is better to get error and fix it.

From what I have seen so far, tensorflow and cupy behave differently regarding this. In tensorflow if a GPU tensor is used in an operation under `with tf.device("/CPU:0")`, it will automatically (and silently) be casted to CPU. In cupy it usually raises an error prompting the user to use `.get()` to explicitly do the GPU to CPU transfer. I am not sure what is the best approach.

Reply:
Ok, then lets keep the version proposed here. Concerning the automatic cast, I also prefer non silent automatic castings, so doing the `.get` on demand looks quite clean to me.